PA193_mnemonic_Slytherin/mnemonic.py

```diff
@@ -38,24 +40,97 @@ def __generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
 # - every line has exactly 1 word (no whitespaces)
 
-def __entropy2mnemonic(entropy: bytes) -> str:
-    """Convert entropy to mnemonic phrase using dictionary.
+def __get_dictionary() -> Tuple[List[str], Dict[str, int]]:
+    """Load the dictionary.
     Currently uses 1 default dictionary with English words.
     # TODO Should we support multiple dictionaries for various languages?
+    :raises FileNotFoundError: on missing file
+    :raises ValueError: on invalid dictionary
+    :rtype: Tuple[List[str], Dict[str, int]]
+    :return: List and dictionary of words
+
+    """
+    l = []
+    with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'english.txt'), 'r') as f:
+        for i in range(2048):
+            l.append(next(f).strip())
+            if len(l[-1]) > 16 or len(l[-1].split()) != 1:
+                raise ValueError('invalid dictionary')
+        try:
+            next(f)
+        except StopIteration:
+            pass
+        else:
+            raise ValueError('invalid dictionary')
+
+    d = {l[i]: i for i in range(len(l))}
+    return (l, d)
+
+
+def __get_entropy_checksum(entropy: bytes, length: int) -> int:
+    """Calculate entropy checksum of set length
+    :rtype: int
+    :return: checksum
+    """
+    entropy_hash = hashlib.sha256(entropy).digest()
+    return int.from_bytes(entropy_hash, byteorder='big') >> 256 - length
+    
+
+def _entropy2mnemonic(entropy: bytes) -> str:
+    """Convert entropy to mnemonic phrase using dictionary.
     :rtype: str
     :return: Mnemonic phrase
+
+    > https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic
+    'Checksum is computed using first ENT/32 bits of SHA256 hash. Concatenated bits are split
+    into groups of 11 bits, each encoding a number from 0-2047, serving as an index into a
+    wordlist.'
     """
-    pass
+    word_list, _ = __get_dictionary()
+    shift = len(entropy) // 4
+    checksum = __get_entropy_checksum(entropy, shift)
 
+    # Concatenated bits representing the indexes
+    indexes_bin = (int.from_bytes(entropy, byteorder='big') << shift) | checksum
 
-def __mnemonic2entropy(mnemonic: str) -> bytes:
+    # List of indexes, number of which is MS = (ENT + CS) / 11 == shift * 3
+    indexes = [(indexes_bin >> i * 11) & 2047 for i in reversed(range(shift * 3))]
+
+    words = [word_list[i] for i in indexes]
+    return ' '.join(words)
+
+
+def _mnemonic2entropy(mnemonic: str) -> bytes:
     """Convert mnemonic phrase to entropy using dictionary.
-    Currently uses 1 default dictionary with English words.
-    # TODO Should we support multiple dictionaries for various languages?
+    :raises ValueError: on invalid parameters
     :rtype: bytes
     :return: Entropy
     """
-    pass
+    _, word_dict = __get_dictionary()
+    words = mnemonic.split()
+    l = len(words)
+    
+    try:
+        indexes = [word_dict[word] for word in words]
```

Comment:
As far as I understand, the check of mnemonics is implemented right inside the function, here and below by calculating the checksum. But what about the allowed length of the mnemonic sentence and, so, the entropy?

Reply:
should be also checked now :)

Reply:
Would be solved by #26. But that would be for a separate PR, don't worry. 😄